MarkupsToModel/Logic/vtkSlicerMarkupsToModelLogic.cxx

```diff
@@ -538,6 +577,373 @@ void vtkSlicerMarkupsToModelLogic::UpdateOutputHermiteSplineModel(vtkMRMLMarkups
+                    << numUniquePointParameters << " points with unique parameters were found. "
+                    << "An order " << (numUniquePointParameters - 1) << " polynomial will be created instead.");
+    numPolynomialCoefficients = numUniquePointParameters;
+  }
+
+  // independent values (parameter along the curve)
+  int numIndependentValues = numPoints * numPolynomialCoefficients;
+  std::vector<double> independentValues(numIndependentValues); // independent values
+  independentValues.assign(numIndependentValues, 0.0);
+  for (int c = 0; c < numPolynomialCoefficients; c++) // o = degree index
+  {
+    for (int p = 0; p < numPoints; p++) // p = point index
+    {
+      double value = std::pow( markupsPointsParameters->GetValue(p), c );
+      independentValues[p * numPolynomialCoefficients + c] = value;
+    }
+  }
+  std::vector<double*> independentMatrix(numPoints);
+  independentMatrix.assign(numPoints, NULL);
+  for (int p = 0; p < numPoints; p++)
+  {
+    independentMatrix[p] = &(independentValues[p * numPolynomialCoefficients]);
+  }
+  double** independentMatrixPtr = &(independentMatrix[0]);
+
+  // dependent values
+  const int numDimensions = 3; // this should never be changed from 3
+  int numDependentValues = numPoints * numDimensions;
+  std::vector<double> dependentValues(numDependentValues); // dependent values
+  dependentValues.assign(numDependentValues, 0.0);
+  for (int p = 0; p < numPoints; p++) // p = point index
+  {
+    double* currentPoint = points->GetPoint(p);
+    for (int d = 0; d < numDimensions; d++) // d = dimension index
+    {
+      double value = currentPoint[d];
+      dependentValues[p * numDimensions + d] = value;
+    }
+  }
+  std::vector<double*> dependentMatrix(numPoints);
+  dependentMatrix.assign(numPoints, NULL);
+  for (int p = 0; p < numPoints; p++)
+  {
+    dependentMatrix[p] = &(dependentValues[p * numDimensions]);
+  }
+  double** dependentMatrixPtr = &(dependentMatrix[0]);
+
+  // solution to least squares
+  int totalNumberCoefficients = numDimensions*numPolynomialCoefficients;
+  std::vector<double> coefficientValues(totalNumberCoefficients);
+  coefficientValues.assign(totalNumberCoefficients, 0.0);
+  std::vector<double*> coefficientMatrix(numPolynomialCoefficients);
+  for (int c = 0; c < numPolynomialCoefficients; c++)
+  {
+    coefficientMatrix[c] = &(coefficientValues[c * numDimensions]);
+  }
+  double** coefficientMatrixPtr = &(coefficientMatrix[0]); // the solution
+
+  // Input the forumulation into SolveLeastSquares
+  vtkMath::SolveLeastSquares(numPoints, independentMatrixPtr, numPolynomialCoefficients, dependentMatrixPtr, numDimensions, coefficientMatrixPtr);
+
+  // Use the values to generate points along the polynomial curve
+  vtkSmartPointer<vtkPoints> smoothedPoints = vtkSmartPointer<vtkPoints>::New(); // points
+  vtkSmartPointer< vtkCellArray > smoothedLines = vtkSmartPointer<  vtkCellArray >::New(); // lines
```

Comment:
Is there an established standard with regards to spaces around < >, [ ], and ( ) brackets? I've seen both used in different code.